pkg/collector/corechecks/snmp/session.go

```diff
@@ -0,0 +1,109 @@
+package snmp
+
+import (
+	"fmt"
+	"github.com/gosnmp/gosnmp"
+	"time"
+)
+
+type sessionAPI interface {
+	Configure(config snmpConfig) error
+	Connect() error
+	Close() error
+	Get(oids []string) (result *gosnmp.SnmpPacket, err error)
+	GetBulk(oids []string) (result *gosnmp.SnmpPacket, err error)
+}
+
+type snmpSession struct {
+	gosnmpInst gosnmp.GoSNMP
+}
+
+func (s *snmpSession) Configure(config snmpConfig) error {
+	maxOids := gosnmp.MaxOids
+	if config.oidBatchSize > gosnmp.MaxOids {
+		return fmt.Errorf("config oidBatchSize (%d) cannot higher than gosnmp.MaxOids: %d", config.oidBatchSize, maxOids)
+	}
+	snmpVersion, err := parseVersion(config.snmpVersion)
+	if err != nil {
+		return err
+	}
+
+	switch snmpVersion {
+	case gosnmp.Version2c, gosnmp.Version1:
+		s.gosnmpInst.Community = config.communityString
+	case gosnmp.Version3:
+		authProtocol, err := getAuthProtocol(config.authProtocol)
+		if err != nil {
+			return err
+		}
+
+		privProtocol, err := getPrivProtocol(config.privProtocol)
+		if err != nil {
+			return err
+		}
+
+		msgFlags := gosnmp.NoAuthNoPriv
+		if privProtocol != gosnmp.NoPriv {
+			msgFlags = gosnmp.AuthPriv
+		} else if authProtocol != gosnmp.NoAuth {
+			msgFlags = gosnmp.AuthNoPriv
+		}
+
+		s.gosnmpInst.MsgFlags = msgFlags
+		s.gosnmpInst.ContextName = config.contextName
+		s.gosnmpInst.SecurityModel = gosnmp.UserSecurityModel
+		s.gosnmpInst.SecurityParameters = &gosnmp.UsmSecurityParameters{
+			UserName:                 config.user,
+			AuthenticationProtocol:   authProtocol,
+			AuthenticationPassphrase: config.authKey,
+			PrivacyProtocol:          privProtocol,
+			PrivacyPassphrase:        config.privKey,
+		}
+	}
+
+	s.gosnmpInst.Target = config.ipAddress
+	s.gosnmpInst.Port = config.port
+	s.gosnmpInst.Version = snmpVersion
+	s.gosnmpInst.Timeout = time.Duration(config.timeout) * time.Second
+	s.gosnmpInst.Retries = config.retries
+	s.gosnmpInst.MaxOids = maxOids
+
+	// Uncomment following line for debugging
+	// s.gosnmpInst.Logger:  defaultLog.New(os.Stdout, "", 0),
```

Comment:
In this case, I would like to avoid exposing gosnmp debug logs to the user. So, checking for DEBUG or current log level is not suitable.

I left this just as a hint in the rare case where the dev need to debug deep down to the library.

Reply:
sorry, meaning the `DEBUG` C++ preprocessor option, that compiles code only when in debug mode - no exposure to users, except if you deliver a debug version.

Reply:
AFAIK, that's not possible in golang :)